File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        if(root == null){
            return 0;
        }else if(isLeaf(root)){
            return 1;
        }else{
            int leftdepth = maxDepth(root.left);
            int rightdepth = maxDepth(root.right);
            return 1+max(leftdepth,rightdepth);
        }
    }

    int max(int a,int b){
        return ((a >= b) ? a : b);
    }

    boolean isLeaf(TreeNode node){
        return (node.left == null && node.right == null);
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        # preorder gives us root
        root = TreeNode(preorder[0])
        # identify root position in inorder to seperate left and right tree using lengths
        indexOfRootinInorder = inorder.index(preorder[0])
        # 0 to midOrder gives us left subtree of root and midOrder+1 to right gives us right subtree
        root.left = self.buildTree(preorder[1:indexOfRootinInorder+1], inorder[0:indexOfRootinInorder])
        root.right = self.buildTree(preorder[indexOfRootinInorder+1:], inorder[indexOfRootinInorder+1:])

        return root